File: advanced-programming-methods/labs/toy-interpreter-java/src/model/expression/HeapNewExpression.java
package model.expression;

import exceptions.ExpressionException;
import model.IDictionary;
import model.IHeap;

public class HeapNewExpression extends Expression
{
    private Expression value;

    public HeapNewExpression(Expression value)
    {
        this.value = value;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int eval(IDictionary table, IHeap heap) throws ExpressionException
    {
        return heap.allocate(value.eval(table, heap));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toString()
    {
        return String.format("new(%s)", value.toString());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Expression cloneDeep()
    {
        return new HeapNewExpression(value.cloneDeep());
    }
}

File: advanced-programming-methods/labs/toy-interpreter-sharp/model/expression/HeapNewExpression.cs
using System;

namespace Model
{
    [Serializable()]
    public class HeapNewExpression : Expression
    {
        private int value;

        public HeapNewExpression(int value)
        {
            this.value = value;
        }

        /// <inheritdoc />
        public override int Eval(IDictionary table, IHeap heap)
        {
            return heap.Allocate(value);
        }

        /// <summary> {@inheritDoc} </summary>
        public override string ToString()
        {
            return string.Format("new({0})", value);
        }

        /// <summary> {@inheritDoc} </summary>
        public override Expression CloneDeep()
        {
            return new HeapNewExpression(value);
        }
    }
}

